src/audit_payloads.py

```diff
@@ -0,0 +1,194 @@
+#	format user@fully.qualified.domain.name
+#     SlotID (integer) - slot identifier number on the worker node
+#     MyType (string) - required to be set to the value of "Machine"
+#   The condor_advertise at the begining of the message must also contain
+#     GlobalJobId (string) - a globally unique identifier of the job
+#     RemoteOwner (string) - a string identifying the owner of the job
+#   and if the beginning of the message contains any of these they will
+#   also be logged:
+#     ClientMachine (string)
+#     ProjectName (string)
+#     Group (string)
+#     x509UserProxyVOName (string)
+#     x509userproxysubject (string)
+#     x509UserProxyFQAN (string)
+#     x509UserProxyEmail (string)
+#
+# There is one condor-ce configuration variable AUDIT_PAYLOAD_MAX_HOURS,
+#   which is optional and indicates the maximum number of hours any job
+#   is expected to run, default 72 hours (3 days).  After that time the
+#   jobs will stop being tracked, in case a stop message was missed.
+#
+# Written by Dave Dykstra, June 2017
+#
+
+import htcondor
+# I would use the simpler "time" module but it gets a load error
+#   undefined symbol: PyExc_ValueError
+from datetime import datetime
+
+maxjobsecs = None
+runningjobs = {}
+firstidx = None
+lastidx = None
+
+def dumpvars():
+    htcondor.log(htcondor.LogLevel.Audit, "firstidx: %s, lastidx: %s, runningjobs: %s" % (firstidx, lastidx, str(runningjobs)))
+
+# stop tracking a running job
+def removerunningjob(idx):
+    global runningjobs
+    global firstidx
+    global lastidx
+
+    if idx not in runningjobs:
+	# shouldn't happen, but just in case
+	return
+
+    # remove this job from the doubly-linked list
+    thisjob = runningjobs[idx]
+    if thisjob['previdx'] == None:
+	firstidx = thisjob['nextidx']
+    else:
+	runningjobs[thisjob['previdx']]['nextidx'] = thisjob['nextidx']
+    if thisjob['nextidx'] == None:
+	lastidx = thisjob['nextidx']
+    else:
+	runningjobs[thisjob['nextidx']]['previdx'] = thisjob['previdx']
+
+    # and remove it from the running jobs
+    del runningjobs[idx]
+
+
+# start tracking a running job
+def addrunningjob(idx, globaljobid, now):
+    global runningjobs
+    global firstidx
+    global lastidx
+
+    if idx in runningjobs:
+	# shouldn't happen, but just in case
+	removerunningjob(idx)
+
+    thisjob = {}
+    thisjob['globaljobid'] = globaljobid
+    thisjob['starttime'] = now
+
+    # append this job the end of the doubly-linked list
+    thisjob['previdx'] =  lastidx
+    thisjob['nextidx'] =  None
+    if lastidx != None:
+	runningjobs[lastidx]['nextidx'] = idx
+    lastidx = idx
+    if firstidx == None:
+	firstidx = idx 
+
+    # and add it to the running jobs
+    runningjobs[idx] = thisjob
+
+# a job may be being stopped
+def stopjob(info):
+    global runningjobs
+    if 'Name' not in info or 'SlotID' not in info:
+	return
+    idx = (info['Name'], info['SlotID'])
+    if idx not in runningjobs:
+	return
```

Comment:
The above 5 lines seem to be mostly duplicated through the module.  Can you refactor them into a helper function?